[Serializer][FrameworkBundle] Add a CSV encoder

dunglas opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | todo |

Usage:

``` php
use Symfony\Component\Serializer\Serializer;
use Symfony\Component\Serializer\Encoder\CsvEncoder;
use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;

$serializer = new Serializer(array(new ObjectNormalizer()), array(new CsvEncoder()));
// or $serializer = $container->get('serializer'); when using the full stack framework
$serializer->encode($something, 'csv');
$serializer->decode(<<<'CSV'
id,name
1,Kévin
CSV
, 'csv');
```

CSV files must contain a header line with property names as keys.

ping @clementtalleu @Simperfit @gorghoa

dunglas committed:
[Serializer] Add a CSV encoder

dunglas committed:
[FrameworkBundle] Register the CSV encoder

Koc commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -977,6 +978,13 @@ private function registerSerializerConfiguration(array $config, ContainerBuilder
             $definition->addTag('serializer.normalizer', array('priority' => -900));
         }
 
+        if (class_exists('Symfony\Component\Serializer\Encoder\CsvEncoder')) {
```
You can use ::class here also

dunglas replied:
It's weird (accessing a constant on a class that doesn't exist) but I confirm it works.

Simperfit commented:
LGTM 👍

gorghoa commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,165 @@
+            $result = array();
+            $this->flatten($value, $result);
+
+            if (!$headersAdded) {
```
@dunglas, instead of this condition, why don’t you just set csv headers before the foreach ?

gorghoa commented:
@dunglas , thanks for your work ;) 

Still, I may be wrong, but I think this case may be problematic with your implementation:

Given an entity (Person) with an attribute (Addresses) with a 0,n association.
Then, Person instances may have an array of  Addresses of different length. 
Hence, the length of the results once filled by the flatten method may be different from one row to another. 
Since you are guessing headers form the first row. Headers and rows may be incoherent.

I know it’s a pretty annoying use case but I think that flattening in this is case may be more a con that a pro. A workaround could be a kind of serialization of the related entities and merge them together in one csv column. Not very elegant but maybe more robust in term of cohesion of the generated file.

Simperfit commented on src/Symfony/Component/Serializer/Tests/Encoder/CsvEncoderTest.php:
```diff
@@ -0,0 +1,155 @@
+
+        $this->assertEquals($expected, $this->encoder->decode(<<<'CSV'
+foo,bar.baz.bat
+a,b,c
```
I don't understand why there `c` here ?

dunglas replied:
To test that it is silently ignored.

Simperfit commented on src/Symfony/Component/Serializer/Tests/Encoder/CsvEncoderTest.php:
```diff
@@ -0,0 +1,155 @@
+        $this->assertEquals($expected, $this->encoder->decode(<<<'CSV'
+foo,bar
+a,b
+c,d,e
```
why is there `e` here ?

dunglas replied:
To test that it is silently ignored.

dunglas replied:
To test that it is silently ignored.

javiereguiluz replied:
Maybe add a PHP comment explaining this ... or changing the method name to `testDecodeMalformedCollection` ?

dunglas commented:
@gorghoa another solution can be to generate something like that:

```csv
id,name,address.0,address.1,address.2,address.3,address.4
1,Kévin,A,B
2,Rodrigue,A,B,C,D,E
```

WDYT?

gorghoa commented:
@dunglas , that may work but I still two flaws:

1st, we’ll need to iterate through all the rows first in order to generate the headers

2nd, If another 0,n attribute (say phones) is present, we can’t use the trick of the last optionable columns (or we have to fill empty columns with null value).

Finally, I think that, with this dynamic set of column, it will be more complex for consumer systems (apart from excel and associates) to handle the csv representation.

javiereguiluz commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,165 @@
+            return $result;
+        }
+
+        // If there is only one data line in the document, return it is not considered has a collection
```
I don't understand well the last part of this phrase: `[...] return it is not considered has a collection`

greg0ire replied:
It should probably be `// If there is only one data line in the document, return it (the line), the result is not considered as a collection` (has => as). Not sure this is very good design : if you can't be sure you can't get one-line long documents, you are going to have to write a special case just for this in client code. I think there should be a dedicated method for this use case, that would throw an exception if more than one line is found.

dunglas replied:
The idea behind this is to allow encoding single items (e.g. `/product/1`) as a CSV document containing 1 one line without having to tune the normalizer. And this is unit tested (`testEncode`/`testEncodeCollection` / `testDecode`/`testDecodeCollection`).

dunglas commented:
It's an implementation detail for 1 and 2 (and I've an idea of how do that, even if it will decrease a bit performances).

IMO this approach is still better - even for naive clients - than generating and output not covered by the RFC (like columns in the column).

dunglas commented:
@gorghoa I've looked at how various projects handle that (Sales Force, Drupal) and they all use a solution similar to yours. So after a thinking again, I'll implement what you suggest.

Probably something like that:

```
id,name,groups
1,Kévin,"a,b,c"
```

javiereguiluz commented:
@dunglas maybe we should also ask for the opinion of Sonata guys (mostly @Soullivaneuh and @greg0ire) because they have experience creating CSV writers (they even have a separate project for that: https://github.com/sonata-project/exporter)

phansys commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,165 @@
+        fwrite($handle, $data);
+        rewind($handle);
+
+        $headers = null;
```
If this var is intended to store an array later, it can be initialized as an empty array IMO. I don't see any special case `null`.

roukmoute replied:
To remove

phansys replied:
I don't get your point @roukmoute. Can you please show where is this happening?

dunglas replied:
I prefer to keep is as is to have a code as close as possible in `encode` and `decode` (and in encode `$headers` cannot be initialized as an empty array because the normalized data can contain an empty array).

roukmoute replied:
@phansys: sorry I write a mistake.

dunglas commented:
IIRC, the Sonata Exporter has no special handling for multiple values. You need to handle yourself by adding a special method to the entity returning a string.

gorghoa commented:
@dunglas:  Can’t argue against the RFC argument ;) I’m ok then with the elastic columns set.

greg0ire commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,165 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function encode($data, $format, array $context = array())
```
What if `$data` is huge ? It would be great to have an interator here instead.

dunglas replied:
It will be an array because of the current design of the Serializer (Normalizers return arrays).

greg0ire commented:
> @dunglas maybe we should also ask for the opinion of Sonata guys (mostly @Soullivaneuh and @greg0ire) because they have experience creating CSV writers (they even have a separate project for that: https://github.com/sonata-project/exporter)

Not speaking for @Soullivaneuh here, as far as I'm concerned, I'm just maintaining these writers, I didn't actually create them. I think a good inspiration source could be https://github.com/ddeboer/data-import/blob/master/src/Writer/CsvWriter.php cc @sagikazarmark @Baachi @ddeboer (input from those guys would be great IMO)

dunglas commented:
There is another issue with the solution proposed by @gorghoa. If it's a collection of complex structures (like a collection of associative arrays), it becomes impossible to display it (or it will require including a full CSV document inside the column, and it looks bad to me). However, the flatten column solution will work:

```csv
name,address.0.line1,address.0.line2,address.1.line1,address.1.line2
```

Thought?

greg0ire commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,165 @@
+        }
+
+        rewind($handle);
+        $value = stream_get_contents($handle);
```
Same problem here, what if `$data` is huge? `$value` will be too. Not sure you can do anything against that though, it depends if Symfony can work with a stream rather than a string…

dunglas replied:
Not currently, but it is impact all other decoders as well. However it will be a nice addition to the component to be able to deal with streams.

greg0ire replied:
It would indeed! Php has `yield` now, which should make this kind of thing easier to work with (What if normalizers yielded values instead of returning a huge array they build ?)

dunglas replied:
It can be done in Symfony 3 because the minimal PHP version is now 5.5. However encoders must stay backward compatible with normalizers returning raw arrays (they must support both cases).

greg0ire replied:
Should I create a new issue about this ?

dunglas replied:
Why not!

greg0ire replied:
adding that to my todo list!

greg0ire replied:
See #19238

GromNaN commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,165 @@
+     * @param string $escapeChar
+     * @param string $keySeparator
+     */
+    public function __construct($delimiter = ',', $enclosure = '"', $escapeChar = '\\', $keySeparator = '.')
```
What about using an array of options instead. That would make the configuration more explicit. This is also make it easier to add options.

greg0ire replied:
And maybe use the option resolver, then?

dunglas replied:
It mimics the prototype of CSV related PHP functions and it carries a strong semantic. Why changing that to (less semantic) array? New options can be added by adding new parameters.

dunglas replied:
Btw we actually this style for all encoders and normalizers. It would be weird to change it for this one.

greg0ire replied:
I think the reasoning behind this is that if you use such an array, you can change the key separator without having to specify the other values. And you can still use the same names for option keys, so… also, I think you will probably add other options not related to prototype later, for instance, a `withBom` boolean.

greg0ire replied:
Here is [how this `withBom` boolean would be used](https://github.com/sonata-project/exporter/blob/1.x/src/Writer/CsvWriter.php#L106). One more argument would be that function parameters are for dependencies, and that configuration should be provided as a separate object / array, because while dependencies are often not so numerous, options can quickly lead you to violate [this rule](https://phpmd.org/rules/codesize.html#excessiveparameterlist)

greg0ire replied:
> Btw we actually use this style for all encoders and normalizers. It would be weird to change it for this one.

But then if you have to be consistent… too bad.

rybakit commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,165 @@
+        $handle = fopen('php://temp,', 'w+');
+
+        // Sequential arrays are considered as collections
+        if (array_keys($data) !== range(0, count($data) - 1)) {
```
I guess this can be simplified to:

```
if (array_values($data) === $data) {
```

dunglas replied:
It is slower.

rybakit replied:
Don't know if it's essential here, but there is a difference between these two checks: https://3v4l.org/XSA3U

rybakit replied:
@dunglas According to my tests, `array_values` is faster (e.g. 2 times faster on 1000 size array) than `array_keys` + `range` + `count`: https://gist.github.com/rybakit/12dc8508b5d144a910bdc2829403d7b0#file-output-txt

rybakit replied:
@dunglas According to my tests, `array_values` is faster (e.g. 2 times faster on 1000 size array) than `array_keys` + `range` + `count`: https://gist.github.com/rybakit/12dc8508b5d144a910bdc2829403d7b0#file-output-txt

dunglas replied:
Thanks for benchmarking this. I tried your solution but as you pointed out, the difference matters. Using `array_values` breaks the tests.

xabbuh replied:
There can still be the case where a simply plain indexed array is passed as `$data` here. It won't be wrapped in the "collection array" and will then break the call to `flatten()` below as `$value` is no array.

dunglas replied:
Fixed, handled the case of non-array too.

greg0ire commented:
> There is another issue with the solution proposed by @gorghoa. If it's a collection of complex structures (like a collection of associative arrays), it becomes impossible to display it (or it will require to include a full CSV document inside the cell, and it looks bad to me). However, the flatten column solution works as expected:

I think CSV is not the best format to represent deep data structures, and that you should simply throw an exception if the input data goes to deep. If people want to flatten things, they can still do it *before* feeding the result to your writer, which should IMO stay focused on doing one thing and doing it well.

dunglas commented:
@greg0ire's idea implemented (It's by far the best idea. Thanks.) and other comments fixed.

gorghoa commented:
Thanks @dunglas, as far as I concerned: it looks good to me ;)

rybakit commented:
@dunglas What about [my comment](https://github.com/symfony/symfony/pull/19197#discussion_r69017729) for the case when `$data = []`? The check in line #50 will do comparison against `range(0, -1)` which evaluates into `[0, -1]`.

greg0ire commented:
> @greg0ire's idea implemented (It's by far the best idea. Thanks.) and other comments fixed.

Thanks! :blush: I'll have a look at it :)

greg0ire commented:
I don't get it… Is it me or is your last push is 4 days old ? :confused: I don't see anything new, but then again github has been having strange caching issues lately.

Simperfit commented:
@greg0ire he rebased, this is why ^^

chalasr commented:
@greg0ire My guess is that the change is [here](https://github.com/symfony/symfony/pull/19197/files#diff-4de6bbbd40ea769ada711de28fb180c8R65).

greg0ire commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,169 @@
+                $headers = array_keys($result);
+                fputcsv($handle, $headers, $this->delimiter, $this->enclosure, $this->escapeChar);
+            } elseif (array_keys($result) !== $headers) {
+                throw new InvalidArgumentException('To use the CSV encoder, each line in the data array must have the same structure. You may want to use a custom normalizer class to normalize the data format before passing it to the CSV encoder.');
```
Maybe linebreak this ?

fabpot replied:
no

greg0ire replied:
Ok, you convinced me !

Soullivaneuh replied:
@fabpot even if I'm personnaly convinced about no linebreak for method and class signatures, I don't really understand why we can't linebreak this thing? It's a new class, no merge conflict possible here.

Could you tell us why? I'm curious. :-) Maybe a coding standard I missed?

I think we can at least put the string argument onto a new line.

Thanks.

fabpot replied:
I like the code to be dense vertically. Nowadays, screens have a lot of horizontal space. And anyway, this is just an error message, I don't want exceptions to take too much screen space when reading the code.

jvasseur replied:
Error message should be keep on one line to allow searching them more easily in the code.

greg0ire replied:
> Nowadays, screens have a lot of horizontal space

That's why we disagree apparently : I use the fact that my screen is big to split windows and often have 4 files on the same screen. I think this is the very first factor of speed on my workflow, because it makes navigating files way easier than browsing through tabs IMO, but I know many people seem to like to use their screen for one file at a time, while they often need to switch between a controller, a service, a view and a translation file in applications, a class and its test in libraries.

Other things to take into consideration is that Github (for instance) will not display this nicely, and also long lines are harder to edit : people scroll way faster vertically than horizontally.

But anyway, your repo, your rules, so thanks for taking the time to give your point of view even if you're probably very busy.

greg0ire replied:
> Error messages should be keep on one line to allow searching them more easily in the code.

That's if you linebreak them in the middle of a sentence.

greg0ire commented:
> @greg0ire he rebased, this is why ^^

Rebasing means you have to push again and rewrite commits, but ok…

jvasseur commented:
@greg0ire but rewriting commits can keep the original commit date, and github use the commit date, not the push date.

greg0ire commented:
> @greg0ire but rewriting commits can keep the original commit date, and github use the commit date, not the push date.

Makes sense :bulb: Thanks!

dunglas commented:
@rybakit 

> @dunglas What about [my comment](https://github.com/symfony/symfony/pull/19197#discussion_r69017729) for the case when `$data = []`? The check in line #50 will do comparison against `range(0, -1)` which evaluates into `[0, -1]`.

I've added 2 unit tests for empty arrays and empty documents (and I fixed a bug in the decoder thanks to this test).

dunglas commented:
Rebased.

ping @symfony/deciders

dunglas commented:
ping @symfony/deciders

xabbuh commented on src/Symfony/Component/Serializer/Encoder/CsvEncoder.php:
```diff
@@ -0,0 +1,169 @@
+     * @param array  $result
+     * @param string $parentKey
+     */
+    private function flatten(array $array, array &$result, $parentKey = '')
```
should be moved after `public` methods

mcfedr commented:
Idea that might be useful - a way of customising the column names. 
In most cases a `Normalizer` would allow you to use custom column names, but for nested columns you will always have the `"."` between the names.

There might be value in being able to pass a mapping of names in the serializer context

An example, you might have had this output:

    name,address.street,address.city
    Fred, A Street, A City

But with a name mapping:

    [
       'address.street' = 'Street',
       'address.city' = 'City
    ]

Changes to 

    name,Street,City
    Fred, A Street, A City

It would be a simple look up when the headers are set, but allow much more flexible CSV generation.

dunglas committed:
Support non-arrays and plain indexed arrays

dunglas committed:
Fix CS

dunglas commented:
@mcfedr good idea but IMO it deserves its own PR. @symfony/deciders are you ok to merge this one?

fabpot commented:
Thank you @dunglas.
